Make analyzer tests fail when target code does not compile

Currently, one can write a test and if the code doesn't compile, the test passes. When I modify the code tested inside `UnusedReturnValueAnalyzerTests.Diagnostic_Should_Not_Be_Triggered_When_Call_Is_Configured_With_Returns` like so:

```c#
using FakeItEasy;
namespace TheNamespace
{
    class TheClass
    {
        void Test()
        {
            var foo = A.Fake<IFoo>();
            A.Call To(() => foo.Bar());
        }
    }

    interface IFoo { int Bar(); }
}
```

(I removed the `Returns`, but accidentally added a space in `A.CallTo`. It's a little contrived, but the sort of problem a copy/paste might introduce)
the test passes.

In some cases, this won't materially affect the quality of the test, but in some it will. 

Instead, we could have the test fail. With output like this:

```
Build failed. Diagnostics are:
  Test0.cs(10,22): error CS1528: Expected ; or = (cannot specify constructor arguments in declaration)
  Test0.cs(10,22): error CS1003: Syntax error, '[' expected
  Test0.cs(10,39): error CS1003: Syntax error, ']' expected
  Test0.cs(10,15): error CS0426: The type name 'Call' does not exist in the type 'A'
  Test0.cs(10,20): warning CS0168: The variable 'To' is declared but never used
```

As you might guess. I have a spike, if we're interested.

--- a/tests/FakeItEasy.Analyzer.Tests/Helpers/DiagnosticVerifier.Helper.cs
+++ b/tests/FakeItEasy.Analyzer.Tests/Helpers/DiagnosticVerifier.Helper.cs
@@ -4,6 +4,7 @@
     using System.Collections.Generic;
     using System.Collections.Immutable;
     using System.Linq;
+    using System.Text;
     using Microsoft.CodeAnalysis;
     using Microsoft.CodeAnalysis.CSharp;
     using Microsoft.CodeAnalysis.Diagnostics;
@@ -53,6 +54,9 @@
             foreach (var project in projects)
             {
                 var compilationWithAnalyzers = project.GetCompilationAsync().Result.WithAnalyzers(ImmutableArray.Create(analyzer));
+
+                AssertThatCompilationSucceeded(compilationWithAnalyzers);
+
                 var diags = compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync().Result;
                 foreach (var diag in diags)
                 {
@@ -115,6 +119,24 @@
             return documents;
         }
 
+        private static void AssertThatCompilationSucceeded(CompilationWithAnalyzers compilationWithAnalyzers)
+        {
+            var compilationDiagnostics = compilationWithAnalyzers.Compilation.GetDiagnostics();
+
+            if (compilationDiagnostics.Any())
+            {
+                var messageBuilder = new StringBuilder();
+                messageBuilder.Append("Test code compilation failed. Error(s) encountered:");
+                foreach (var diagnostic in compilationDiagnostics)
+                {
+                    messageBuilder.AppendLine();
+                    messageBuilder.AppendFormat("  {0}", diagnostic);
+                }
+
+                throw new ArgumentException(messageBuilder.ToString());
+            }
+        }
+
         /// <summary>
         /// Create a project using the inputted strings as sources.
         /// </summary>
@@ -131,6 +153,7 @@
             var solution = new AdhocWorkspace()
                 .CurrentSolution
                 .AddProject(projectId, TestProjectName, TestProjectName, language)
+                .WithProjectCompilationOptions(projectId, new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary))
                 .AddMetadataReference(projectId, CorlibReference)
                 .AddMetadataReference(projectId, SystemCoreReference)
                 .AddMetadataReference(projectId, CSharpSymbolsReference)

--- a/tests/FakeItEasy.Analyzer.Tests/UnusedReturnValueAnalyzerTests.cs
+++ b/tests/FakeItEasy.Analyzer.Tests/UnusedReturnValueAnalyzerTests.cs
@@ -84,7 +84,7 @@ namespace TheNamespace
 {
     class TheClass
     {
-        IVoidConfiguration Test()
+        IReturnValueArgumentValidationConfiguration<int> Test()
         {
             var foo = A.Fake<IFoo>();
             return A.CallTo(() => foo.Bar());
@@ -341,7 +341,7 @@ namespace TheNamespace
         void Test()
         {
             var foo = A.Fake<IFoo>();
-            A.CallToSet(() => foo.Bar).To(() => A<int>.That.Matches(i => i > 3);
+            A.CallToSet(() => foo.Bar).To(() => A<int>.That.Matches(i => i > 3));
         }
     }
 
@@ -355,7 +355,7 @@ namespace TheNamespace
                 {
                     Id = DiagnosticDefinitions.UnusedCallSpecification.Id,
                     Message =
-                        "Unused call specification 'A.CallToSet(() => foo.Bar).To(() => A<int>.That.Matches(i => i > 3)'; did you forget to configure or assert the call?",
+                        "Unused call specification 'A.CallToSet(() => foo.Bar).To(() => A<int>.That.Matches(i => i > 3))'; did you forget to configure or assert the call?",
                     Severity = DiagnosticSeverity.Error,
                     Locations = new[] { new DiagnosticResultLocation("Test0.cs", 9, 13) }
                 });

